### PR TITLE
Add support of tty raw mode

### LINKED ISSUE #1531: Change watch listening mode to raw
ava@0.22

It would be nice to change watch listening mode to raw so that users don't have to type `r + ENTER` or `u + ENTER`. Just `r` or `u` is sufficient.

I can work on a PR for that.

However, I notice the code also check for `rs`: https://github.com/avajs/ava/blob/master/lib/watcher.js#L278
which does the same thing as `r`

Any reason for that? if not, I can remove it (which simplifies the code).

### PR DESCRIPTION
**Pull request by unional**
Fixes #1531 

This PR mostly works except killing the process requires hitting [Ctrl+C] twice, which guess is related to the test is running in a different process.

Let me know what is the right way to fix it and I'll adjust the PR. 🌷

### CODE REVIEW

**Comment by novemberborn** [comment]
> This PR mostly works except killing the process requires hitting [Ctrl+C] twice, which guess is related to the test is running in a different process.

I think this is what made us revert https://github.com/avajs/ava/pull/658. How does Jest handle this?

**Comment by unional** [comment]
Jest just `process.exit(0)`
https://github.com/facebook/jest/blob/9c429cd15d639ab42cfe0a9fa7032095af3b675f/packages/jest-cli/src/watch.js#L143
My guess is that they put the watcher in the main process?

I add this PR as a prelude of #1530, so that user can press 'p' and then type in the filter string and 'Enter' to execute.

**Comment by unional** [comment]
Once we figure out how to solve the double [Ctrl-C], I can change this PR to use `keypress` similar to #658 to get the code coverage up.

File: lib/cli.js [hunk 1]
```diff
@@ -182,16 +182,11 @@ exports.run = () => {
 	if (conf.watch) {
 		try {
 			const watcher = new Watcher(logger, api, files, arrify(conf.source));
-			watcher.observeStdin(process.stdin);
+			watcher.observeStdin(process.stdin).then(() => {
+				logger.exit(0);
+			}).catch(handleWatchError);
```
Comment (novemberborn) [inline lib/cli.js]: I don't quite see how the new `observeStdin` would throw this error where the old one doesn't. Why does it need to be handled here?

File: lib/cli.js [hunk 2]
```diff
@@ -182,16 +182,11 @@ exports.run = () => {
 	if (conf.watch) {
 		try {
 			const watcher = new Watcher(logger, api, files, arrify(conf.source));
-			watcher.observeStdin(process.stdin);
+			watcher.observeStdin(process.stdin).then(() => {
+				logger.exit(0);
```
Comment (novemberborn) [inline lib/cli.js]: Instead of exiting directly, can we stop all workers and let the process exit on its own?

**Comment by unional** [comment]
> To clarify, we're blocked until that's resolved, yes?

Yes. Need to figure out how to handle that. In `jest`, it just `process.exit(0)`. Is there some special handling in `ava`?

File: lib/cli.js (hunk 2 above)
Comment (unional) [reply lib/cli.js]: Sure. How to stop those workers?
By the way, I saw in `test-worker.js`:
```js
process.on('ava-exit', () => { process.exit(0); })
```
So the watcher should fire that even in all sub-processes?

File: lib/cli.js (hunk 1 above)
Comment (unional) [reply lib/cli.js]: Yes. The code still needs some love. `observeStdin` is returning a promise that is not rejecting.
Will fix that.

File: lib/cli.js (hunk 2 above)
Comment (novemberborn) [reply lib/cli.js]: Not quite. The child processes should be [sent the `SIGINT` signal](https://nodejs.org/api/child_process.html#child_process_subprocess_kill_signal). See https://github.com/avajs/ava/blob/bb91862eaac1a3e89f651afb15ec9691745ee222/lib/fork.js.

**Comment by novemberborn** [comment]
Hi @unional, I'm sorry to see that this is closed. Did you come to the conclusion that this isn't possible the way we currently want it, or did you close it because your otherwise busy, or perhaps for another reason?

**Comment by unional** [comment]
I think this can be better addressed when the code is refactored. I'll give it a try on that route first before getting back to this. 🌷

**Pull request closed without merging**